backend/cfg/cfg_regalloc_validate.ml

```diff
@@ -0,0 +1,1073 @@
+      t
+
+  let remove_result ~reg_res ~loc_res t =
+    try
+      Array.iter2 (fun reg loc -> compatible_one ~reg ~loc t) reg_res loc_res;
+      let t =
+        array_fold2 (fun t reg loc -> remove (reg, loc) t) t reg_res loc_res
+      in
+      Ok t
+    with Verification_failed message -> Error message
+
+  let verify_destroyed_locations ~destroyed t =
+    (* CR azewierzejew for azewierzejew: Add checking stack for stack_location
+       other than Local. *)
+    try
+      Array.iter
+        (fun destroyed_loc ->
+          iter
+            (fun (live_reg, live_loc) ->
+              if destroyed_loc = live_loc
+              then (
+                Format.fprintf Format.str_formatter
+                  "Destroying a location %a in which a live register %a is \
+                   stored"
+                  Location.print live_loc Register.print live_reg;
+                let message = Format.flush_str_formatter () in
+                raise (Verification_failed message)))
+            t)
+        destroyed;
+      Ok t
+    with Verification_failed message -> Error message
+
+  let add_argument ~reg_arg ~loc_arg t =
+    array_fold2 (fun t reg loc -> add (reg, loc) t) t reg_arg loc_arg
+
+  let rename_loc ~arg ~res t =
+    map
+      (fun ((stamp, loc) as eq) ->
+        if Location.equal loc res then stamp, arg else eq)
+      t
+
+  let rename_reg ~arg ~res t =
+    map
+      (fun ((eq_reg, loc) as eq) ->
+        if Register.equal eq_reg res then arg, loc else eq)
+      t
+
+  let print ppf t =
+    let first = ref true in
+    iter
+      (fun eq ->
+        if !first then first := false else Format.fprintf ppf " ";
+        Format.fprintf ppf "%a" Equation.print eq)
+      t
+end
+
+let extract_loc_arr loc_arr = Array.map Location.of_reg_exn loc_arr
+
+module type Description_value = sig
+  val description : Description.t
+end
+
+let print_reg_as_loc ppf reg =
+  Printmach.loc ~reg_class:(Proc.register_class reg)
+    ~unknown:(fun ppf -> Format.fprintf ppf "<Unknown>")
+    ppf reg.Reg.loc
+
+module Domain : sig
+  (** This type corresponds to the domain of the dataflow from the paper [1]. It
+      is represented as [(Ok equations)] corresponing to the set of equations
+      from the paper [1] or [(Error error)] which corresponds to the "top" from
+      the paper [1]. For the purposes of dataflow all [(Error error)] values are
+      equal even though the carried error can be different. That is because we
+      want to get an error information but if there are multiple errors we do
+      not care which one we see. *)
+  include Cfg_dataflow.Backward_domain
+
+  module Error : sig
+    type t
+
+    module Tag : sig
+      type 'a t =
+        | Terminator : terminator t
+        | Basic : basic t
+    end
+
+    val print : Format.formatter -> t -> unit
+  end
+
+  (** This corresponds to case (10) in Fig. 1 of the paper [1]. *)
+  val rename_location : t -> loc_instr:_ instruction -> t
```

Comment:
This and the following two functions should be part of the transformer, not the domain. Can the Domain.t be not abstract (but immutable), so that the trasformer can operate on it?

Reply:
Moved them.